### ISSUE TITLE
6.0.0 requires changes to some pre_load hooks

### ISSUE THREAD

**Issue by Bangertm** [marshmallow-code/webargs#485]
Just upgraded to 6.0.0. Using `FlaskParser`, the input to `Schema.load` is now a `MultiDictProxy` wrapping a werkzeug `ImmutableMultiDict` instead of a regular dictionary. This breaks a schema I have that uses a `pre_load` hook to converts empty strings to `None`. (I'm processing an HTML form with a bunch of optional number inputs similar to some of the solutions in marshmallow-code/marshmallow#713)

Not the biggest deal to work around but I thought I would mention it because I didn't see it in the change log.

Also it would be convenient if `MultiDictProxy` had a `__repr__` to make debugging more convenient.

**Comment by lafrech** [comment]
Thanks for bringing this up.

Do you have an example of such a change (before/after webargs 6 pre-processor)?

Good point abour `__repr__`.

**Comment by Bangertm** [comment]
The following works with webargs < 6.0.0:

```python
from flask import Flask
from marshmallow import Schema, fields, pre_load
from webargs.flaskparser import use_kwargs

class DemoSchema(Schema):
    number = fields.Number(missing=None)

    @pre_load
    def convert_empty_string_to_none(self, data):
        for key, value in data.items():
            if value == "":
                data[key] = None
        return data

app = Flask(__name__)

@app.route("/", methods=["Post"])
@use_kwargs(DemoSchema(strict=True))
def demo(number):
    print(number)
    return {}

app.test_client().post("/", data={"number": ""})
```

With 6.0.0 and the following change: 
```python
@use_kwargs(DemoSchema(strict=True), location="form")
```

It throws 
```console
TypeError: 'ImmutableMultiDict' objects are immutable
```
when attempting to set `data[key]`.

**Comment by sirosen** [comment]
@Bangertm, are you sure that your `pre_load` hook was even running before 6.0? I'm pretty sure that on 5.x, the hook would never even get run. If your view is getting `number=None` in your example, maybe something else is happening to do that.
It looks to me like that `TypeError` is coming from a werkzeug `ImmutableMultiDict` from flask -- we proxify the `__setitem__` method, but when it runs and tries to set a value, the flask/werkzeug object rejects the operation.
I'll play with your example on the 5.x line, hopefully this evening, before assuming that I'm right on either count, but I think that's what's going on.

@lafrech, maybe we should add some specific note to the changelog about the fact that schema-level hooks like `pre_load`, `post_load`, and `validates_schema` did not run under 5.x and this has been changed/fixed in 6.0 ?

As for the repr, I definitely agree we should add one. I'd probably write it like
```python
def __str__(self):  # str(proxy) proxies to str(proxy.data)
    return str(self.data)
def __repr__(self):
    return "MultiDictProxy(data={!r}, multiple_keys={!r})".format(self.data, self.multiple_keys)
```

unless someone has a better idea for what it should be.
I hope to get to this soon, but we'll see what my schedule allows.

**Comment by Bangertm** [comment]
@sirosen, the `pre_load` hook is definitely being run with pre 6.0. You can verify this by just deleting the hook in the code above. The result is a 422 status returned with the view function not being called. 

I think the issue with `pre_load` hooks pre 6.0 is that they weren't used when gathering parameters from various locations as described in #173. The `pre_load` hook is still called but just called with the wrong data.

For my case the fix was pretty simple - just make a new dict. We may want to consider some more generic commentary on using a `MultiDictProxy` with `pre_load` hooks.

**Comment by sirosen** [comment]
> I think the issue with `pre_load` hooks pre 6.0 is that they weren't used when gathering parameters from various locations as described in #173. The `pre_load` hook is still called but just called with the wrong data.

Yes, you're spot on. I had gotten myself mixed up; sorry about that.
In 5.x, parsing assembled a new dict of data, which is what got passed to the `schema.load` method (and thereby, to `pre_load` and `post_load`).

That also explains why your hook worked in 5.x, but not in 6.0. In 5.x, you get a `dict` with your parameters. But the `MultiDictProxy` you get in 6.0 wraps the raw flask/werkzeug datatype (an `ImmutableMultiDict`).

I just looked, and [werkzeug supports converting an ImmutableMultiDict to a (mutable) MultiDict](https://github.com/pallets/werkzeug/blob/d6e98a0105ea126f10c432d33f101ec793df6440/src/werkzeug/datastructures.py#L1691).
So I think for right now you could start your `pre_load` hook with a line like this
```python
    @pre_load
    def convert_empty_string_to_none(self, data):
        # unfortunate names... makes the underlying datatype mutable
        data.data = data.data.copy()
```

I'm thinking about whether or not / how webargs could support this more "out of the box". For example, should we just call `copy()` on all flask multidicts on their way into the proxy object? (We'd have to check out each framework on this front too.) I don't want to rush out a solution, but this seems like a usage we need to support -- rewriting the contents of the dict in a `pre_load` hook.

**Comment by Bangertm** [comment]
It seems like I can just call `dict` on the `MultiDictProxy` object. I _think_ this works generally:

```python
from werkzeug.datastructures import ImmutableMultiDict
from marshmallow import Schema, fields
from webargs.multidictproxy import MultiDictProxy

class MySchema(Schema):
    a = fields.List(fields.Str())

d = ImmutableMultiDict([('a', 'b'), ('a', 'c')])

mdp = MultiDictProxy(d, MySchema())

print(dict(mdp))
```

That prints the dict: `{'a': ['b', 'c']}`, which looks like what we want - standard dictionary formatted to work with Marshmallow lists. The benefit of this over over pre 6.0 is that if other keys are in the original `ImmutableMultiDict` they will also be in the new dictionary. That allows the enveloping use case to work.

**Comment by sirosen** [comment]
In #489 we're documenting the required changes, but I want to acknowledge that there's an idea here which we could explore further.

The above comment notes that `dict(MultiDictProxy(...))` seems to give the right results for passing to the schema.
I think that will always be true, and, if it works, it improves the interface for users.

However, I'm not sure we'd consider that safe to include in a `6.1.0` release. I think changing that -- even if it's better for everyone -- might have to wait for `7.0`.
I'm going to open a follow-up issue to document this as a potential change and ask if it should be done in 6.1 or 7.0.

**Comment by Bangertm** [comment]
I'm not 100% sure converting to a standard dict is necessarily the right thing for webargs. There's probably a small performance penalty to always converting to a dict instead of just allowing a user to do it in the `pre_load` hook.

If I had seen the note in the upgrade doc I would have been completely fine. 

Going forward it might be good to include a short note in the main documentation maybe under [marshmallow Integration](https://github.com/marshmallow-code/webargs/blob/dev/docs/advanced.rst#L90) like "Working with pre-load hooks" because this would likely come up in the future. For example the first example in [Extending Schemas docs](https://marshmallow.readthedocs.io/en/stable/extending.html) wouldn't work with webargs.

**Comment by sirosen** [comment]
If you're comfortable with us asking users (like yourself) to do the conversion, I'm happy with that. The performance penalty is probably negligible in the majority of cases, but I'm sure that for some combination of location/framework it would be an issue. It would be great to avoid it.

Maybe I should just close #491 .
I can try to add something reasonable to the docs soon, once the upgrade doc is finished. The location you suggested looks good to me.

**Comment by reinhard-mueller** [comment]
If I may join the discussion here, I actually agree with @sirosen's previously stated view that generally converting to a dict improves the user interface and makes the library more intuitively usable. I don't think that the performance penalty is relevant at all, since we're typically not talking about dicts with hundreds of keys.
